lib/Doctrine/Annotations/Metadata/MetadataCollection.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations\Metadata;
+
+use ArrayAccess;
+use Countable;
+use IteratorAggregate;
+use Traversable;
+use function array_key_exists;
+use function array_values;
+use function assert;
+use function count;
+use function sprintf;
+
+/**
+ * @internal
+ */
+final class MetadataCollection implements ArrayAccess, Countable, IteratorAggregate
+{
+    /** @var array<string, AnnotationMetadata> */
+    private $metadata = [];
+
+    public function __construct(AnnotationMetadata ...$metadatas)
+    {
+        $this->add(...$metadatas);
+    }
+
+    public function add(AnnotationMetadata ...$metadatas) : void
+    {
+        foreach ($metadatas as $metadata) {
+            assert(! isset($this[$metadata->getName()]), sprintf('Metadata with name %s already exists.', $metadata->getName()));
```

Comment:
One more assertion that should be an exception: you really don't want this check to be ignored.